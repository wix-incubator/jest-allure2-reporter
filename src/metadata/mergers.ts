import type {
  AllureTestItemMetadata,
  AllureTestFileMetadata,
  AllureTestCaseMetadata,
  AllureTestStepMetadata,
} from 'jest-allure2-reporter';

export function mergeTestFileMetadata(
  a: AllureTestFileMetadata,
  b: AllureTestFileMetadata | undefined,
): AllureTestFileMetadata {
  return b
    ? {
        ...mergeTestItemMetadata(a, b),
        code: undefined,
        steps: undefined,
        workerId: b.workerId ?? a.workerId,
      }
    : a;
}

export function mergeTestCaseMetadata(
  a: AllureTestCaseMetadata,
  b: AllureTestCaseMetadata | undefined,
): AllureTestCaseMetadata {
  return b
    ? {
        ...mergeTestItemMetadata(a, b),
        descriptionHtml: mergeArrays(a.descriptionHtml, b.descriptionHtml),
        historyId: b.historyId ?? a.historyId,
        labels: mergeArrays(a.labels, b.labels),
        links: mergeArrays(a.links, b.links),
      }
    : a;
}

export function mergeTestStepMetadata(
  a: AllureTestStepMetadata,
  b: AllureTestStepMetadata | undefined,
): AllureTestStepMetadata {
  return b
    ? {
        ...mergeTestItemMetadata(a, b),
        hookType: b.hookType ?? a.hookType,
      }
    : a;
}

function mergeTestItemMetadata(
  a: AllureTestItemMetadata,
  b: AllureTestItemMetadata | undefined,
): AllureTestItemMetadata {
  return b
    ? {
        attachments: mergeArrays(a.attachments, b.attachments),
        currentStep: b.currentStep ?? a.currentStep,
        sourceCode: b.sourceCode ?? a.sourceCode,
        sourceLocation: b.sourceLocation ?? a.sourceLocation,
        description: mergeArrays(a.description, b.description),
        parameters: mergeArrays(a.parameters, b.parameters),
        stage: b.stage ?? a.stage,
        start: b.start ?? a.start,
        status: b.status ?? a.status,
        statusDetails: b.statusDetails ?? a.statusDetails,
        steps: b.steps ?? a.steps,
        stop: b.stop ?? a.stop,
      }
    : a;
}

function mergeArrays<T>(
  a: T[] | undefined,
  b: T[] | undefined,
): T[] | undefined {
  if (a && b) {
    return [...a, ...b];
  }

  return a ?? b;
}
