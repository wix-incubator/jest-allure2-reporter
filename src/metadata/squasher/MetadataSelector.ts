import type { AllureTestItemMetadata, AllureTestStepMetadata } from 'jest-allure2-reporter';

import { weakMemoize } from '../../utils';

export type MetadataSelectorOptions<Metadata, Result> = {
  empty(): Result;
  getMetadata(metadata: Metadata): Result | undefined;
  getDocblock(metadata: Metadata): Result | undefined;
  mergeUnsafe(a: Result, b: Result | undefined): Result;
};

export interface HasParent<Metadata> {
  readonly parent?: LikeDescribeBlock<Metadata>;
}

export type LikeDescribeBlock<Metadata> = Metadata & HasParent<Metadata>;

export type LikeTestDefinition<Metadata> = Metadata & {
  readonly describeBlock: LikeDescribeBlock<Metadata>;
};

export type LikeStepInvocation<Metadata> = Metadata & {
  readonly definition: Metadata;
};

export type LikeTestFile<Metadata> = Metadata & {
  readonly globalMetadata: Metadata;
};

export type LikeTestInvocation<Metadata> = Metadata & {
  allInvocations(): Iterable<Metadata>;
  readonly file: LikeTestFile<Metadata>;
  readonly beforeAll: Iterable<LikeStepInvocation<Metadata>>;
  readonly beforeEach: Iterable<LikeStepInvocation<Metadata>>;
  readonly afterEach: Iterable<LikeStepInvocation<Metadata>>;
  readonly afterAll: Iterable<LikeStepInvocation<Metadata>>;
  readonly definition: LikeTestDefinition<Metadata>;
  readonly fn?: Metadata;
};

export class MetadataSelector<Metadata, T extends AllureTestItemMetadata> {
  constructor(protected readonly _options: MetadataSelectorOptions<Metadata, T>) {
    this.belowTestInvocation = weakMemoize(this.belowTestInvocation.bind(this));
    this.testInvocationAndBelow = weakMemoize(this.testInvocationAndBelow.bind(this));
    this.testInvocationAndBelowDirect = weakMemoize(this.testInvocationAndBelowDirect.bind(this));
    this.testDefinitionAndBelow = weakMemoize(this.testDefinitionAndBelow.bind(this));
  }

  protected readonly _getDocblock = (metadata: Metadata | undefined): T | undefined =>
    metadata ? this._options.getDocblock(metadata) : undefined;

  protected readonly _getMetadataUnsafe = (metadata: Metadata | undefined): T | undefined =>
    metadata ? this._options.getMetadata(metadata) : undefined;

  readonly getMetadataWithDocblock = (metadata: Metadata | undefined): T =>
    this._options.mergeUnsafe(
      this._options.mergeUnsafe(this._options.empty(), this._getDocblock(metadata)),
      this._getMetadataUnsafe(metadata),
    );

  protected _describeAncestors = weakMemoize(
    (metadata: LikeDescribeBlock<Metadata> | undefined): T =>
      metadata
        ? this._options.mergeUnsafe(
            this._describeAncestors(metadata.parent),
            this._getMetadataUnsafe(metadata),
          )
        : this._options.empty(),
  );

  protected _ancestors(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(
      this.globalAndTestFile(metadata.file),
      this._describeAncestors(metadata.definition.describeBlock),
    );
  }

  protected _hookMetadata = (metadata: LikeStepInvocation<Metadata>): T => {
    const definitionMetadata = this.getMetadataWithDocblock(metadata.definition);
    return this.merge(definitionMetadata, this._getMetadataUnsafe(metadata));
  };

  public readonly merge = (a: T | undefined, b: T | undefined): T =>
    this._options.mergeUnsafe(this._options.mergeUnsafe(this._options.empty(), a), b);

  testInvocation(metadata: Metadata): T {
    return this.getMetadataWithDocblock(metadata);
  }

  testDefinition(metadata: LikeTestInvocation<Metadata>): T {
    return this.getMetadataWithDocblock(metadata.definition);
  }

  belowTestInvocation(metadata: LikeTestInvocation<Metadata>): T {
    return [...metadata.allInvocations()]
      .map(this._getMetadataUnsafe)
      .reduce(this._options.mergeUnsafe, this._options.empty());
  }

  testInvocationAndBelow(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(this.testInvocation(metadata), this.belowTestInvocation(metadata));
  }

  testInvocationAndBelowDirect(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(this.testInvocation(metadata), this.getMetadataWithDocblock(metadata.fn));
  }

  testDefinitionAndBelow(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(this.testDefinition(metadata), this.testInvocationAndBelow(metadata));
  }

  testDefinitionAndBelowDirect(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(this.testDefinition(metadata), this.testInvocationAndBelowDirect(metadata));
  }

  testVertical(metadata: LikeTestInvocation<Metadata>): T {
    const ancestors = this._ancestors(metadata);
    const test_definition_and_below = this.testDefinitionAndBelow(metadata);
    return this.merge(ancestors, test_definition_and_below);
  }

  steps(metadata: LikeTestInvocation<Metadata>): AllureTestStepMetadata[] {
    const before = [...metadata.beforeAll, ...metadata.beforeEach].map(this._hookMetadata);

    const after = [...metadata.afterEach, ...metadata.afterAll].map(this._hookMetadata);

    const testFunctionSteps =
      (metadata.fn && this.getMetadataWithDocblock(metadata.fn).steps) || [];

    return [...before, ...testFunctionSteps, ...after];
  }

  globalAndTestFile(metadata: LikeTestFile<Metadata>): T {
    return this.merge(
      this._getMetadataUnsafe(metadata.globalMetadata),
      this.getMetadataWithDocblock(metadata),
    );
  }

  globalAndTestFileAndTestInvocation(metadata: LikeTestInvocation<Metadata>): T {
    return this.merge(this.globalAndTestFile(metadata.file), this._getMetadataUnsafe(metadata));
  }
}
