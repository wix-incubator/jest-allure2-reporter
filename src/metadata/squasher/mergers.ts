import type {
  AllureTestItemMetadata,
  AllureTestFileMetadata,
  AllureTestCaseMetadata,
  AllureTestStepMetadata,
  AllureTestStepPath,
  Stage,
  Status,
  StatusDetails,
} from 'jest-allure2-reporter';

export function mergeTestFileMetadata(
  a: AllureTestFileMetadata,
  b: AllureTestFileMetadata | undefined,
): AllureTestFileMetadata {
  return mergeTestCaseMetadata(a, b);
}

export function mergeTestCaseMetadata(
  a: AllureTestCaseMetadata,
  b: AllureTestCaseMetadata | undefined,
): AllureTestCaseMetadata {
  return b
    ? {
        ...mergeTestItemMetadata(a, b),
        description: mergeArrays(a.description, b.description),
        descriptionHtml: mergeArrays(a.descriptionHtml, b.descriptionHtml),
        historyId: b.historyId ?? a.historyId,
        labels: mergeArrays(a.labels, b.labels),
        links: mergeArrays(a.links, b.links),
        workerId: b.workerId ?? a.workerId,
      }
    : a;
}

export function mergeTestStepMetadata(
  a: AllureTestStepMetadata,
  b: AllureTestStepMetadata | undefined,
): AllureTestStepMetadata {
  return b
    ? {
        ...mergeTestItemMetadata(a, b),
        hookType: b.hookType ?? a.hookType,
        steps: mergeArrays(a.steps, b.steps)?.map((step) =>
          mergeTestStepMetadata({}, step),
        ),
      }
    : a;
}

function mergeTestItemMetadata(
  a: AllureTestItemMetadata,
  b: AllureTestItemMetadata | undefined,
): AllureTestItemMetadata {
  return b
    ? {
        attachments: mergeArrays(a.attachments, b.attachments),
        currentStep: mergeCurrentStep(a, b),
        displayName: b.displayName ?? a.displayName,
        sourceCode: b.sourceCode ?? a.sourceCode,
        sourceLocation: b.sourceLocation ?? a.sourceLocation,
        parameters: mergeArrays(a.parameters, b.parameters),
        stage: mergeStage(b.stage, a.stage),
        start: min(b.start, a.start),
        status: mergeStatus(a.status, b.status),
        statusDetails: mergeStatusDetails(a, b),
        stop: max(b.stop, a.stop),
      }
    : a;
}

function mergeArrays<T>(
  a: T[] | undefined,
  b: T[] | undefined,
): T[] | undefined {
  if (a && b) {
    return [...a, ...b];
  }

  return a ?? b;
}

function min(a: number | undefined, b: number | undefined): number | undefined {
  if (a != null && b != null) {
    return Math.min(a, b);
  }

  return a ?? b;
}

function max(a: number | undefined, b: number | undefined): number | undefined {
  if (a != null && b != null) {
    return Math.max(a, b);
  }

  return a ?? b;
}

function mergeStage(
  a: Stage | undefined,
  b: Stage | undefined,
): Stage | undefined {
  if (a === 'interrupted' || b === 'interrupted') {
    return 'interrupted';
  }

  return b ?? a;
}

function mergeStatus(
  a: Status | undefined,
  b: Status | undefined,
): Status | undefined {
  if (a === 'broken' || b === 'broken') {
    return 'broken';
  }

  if (a === 'failed' || b === 'failed') {
    return 'failed';
  }

  return b ?? a;
}

function mergeStatusDetails(
  a: AllureTestItemMetadata,
  b: AllureTestItemMetadata,
): StatusDetails | undefined {
  if (a.status === 'broken') {
    return a.statusDetails;
  }

  if (b.status === 'broken') {
    return b.statusDetails;
  }

  if (a.status === 'failed') {
    return a.statusDetails;
  }

  if (b.status === 'failed') {
    return b.statusDetails;
  }

  return b.statusDetails ?? a.statusDetails;
}

function mergeCurrentStep(
  a: AllureTestItemMetadata,
  b: AllureTestItemMetadata,
): AllureTestStepPath | undefined {
  if (a.stage === 'interrupted') {
    return a.currentStep;
  }

  if (b.stage === 'interrupted') {
    return b.currentStep;
  }

  return b.currentStep ?? a.currentStep;
}
