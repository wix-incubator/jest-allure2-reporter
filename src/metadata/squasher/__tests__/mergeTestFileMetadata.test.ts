import type {
  AllureTestFileMetadata,
  AllureTestStepMetadata,
} from 'jest-allure2-reporter';

import { MetadataSelector } from '../MetadataSelector';
import { mergeTestFileMetadata, mergeTestStepMetadata } from '../mergers';

import {
  createTestFileMetadata,
  createTestStepMetadata,
  getFullBlownTestCase,
  type StubTestFileMetadata,
  type StubTestStepMetadata,
} from './fixtures';

describe('mergeTestFileMetadata', () => {
  let docblocks = new WeakMap();
  let stepSelector: MetadataSelector<
    StubTestStepMetadata,
    AllureTestStepMetadata
  >;
  let fileSelector: MetadataSelector<
    StubTestFileMetadata,
    AllureTestFileMetadata
  >;
  let testCase: ReturnType<typeof getFullBlownTestCase>;

  beforeEach(() => {
    docblocks = new WeakMap();
  });

  beforeEach(() => {
    fileSelector = new MetadataSelector({
      empty: () => ({}),
      getDocblock: (metadata) => docblocks.get(metadata),
      getMetadata: (metadata) => metadata.data,
      mergeUnsafe: mergeTestFileMetadata,
    });

    stepSelector = new MetadataSelector({
      empty: () => ({}),
      getDocblock: (metadata) => docblocks.get(metadata),
      getMetadata: (metadata) => metadata.data,
      mergeUnsafe: mergeTestStepMetadata,
    });

    testCase = getFullBlownTestCase();
    testCase.file.data.steps = [
      createTestStepMetadata('step_invocation', { steps: [{}] }),
    ];
    docblocks.set(testCase.file, createTestFileMetadata('file_docblock'));
  });

  test('merge', () => {
    expect(fileSelector.merge(void 0, void 0)).toMatchSnapshot();
    expect(fileSelector.merge({}, {})).toMatchSnapshot();
  });

  test('getMetadataWithDocblock', () => {
    expect(
      fileSelector.getMetadataWithDocblock(testCase.file),
    ).toMatchSnapshot();
  });

  test('globalAndTestFile', () => {
    expect(fileSelector.globalAndTestFile(testCase.file)).toMatchSnapshot();
  });

  test('steps', () => {
    expect(
      stepSelector.getMetadataWithDocblock(testCase.file).steps,
    ).toMatchSnapshot();
  });
});
