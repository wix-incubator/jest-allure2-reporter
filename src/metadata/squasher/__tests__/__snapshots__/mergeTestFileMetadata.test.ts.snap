// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mergeTestFileMetadata getMetadataWithDocblock 1`] = `
{
  "attachments": [
    {
      "name": "file_docblock.txt",
      "source": "/tmp/file_docblock.txt",
      "type": "text/plain",
    },
    {
      "name": "file.txt",
      "source": "/tmp/file.txt",
      "type": "text/plain",
    },
  ],
  "currentStep": undefined,
  "description": [
    "description:file_docblock",
    "description:file",
  ],
  "descriptionHtml": [
    "descriptionHtml:file_docblock",
    "descriptionHtml:file",
  ],
  "displayName": "displayName:file",
  "fullName": "fullName:file",
  "historyId": "historyId:file",
  "labels": [
    {
      "name": "tag",
      "value": "file_docblock",
    },
    {
      "name": "tag",
      "value": "file",
    },
  ],
  "links": [
    {
      "name": "link",
      "type": "file_docblock",
      "url": "https://example.com/file_docblock",
    },
    {
      "name": "link",
      "type": "file",
      "url": "https://example.com/file",
    },
  ],
  "parameters": [
    {
      "name": "parameter:file_docblock",
      "value": "value:file_docblock",
    },
    {
      "name": "parameter:file",
      "value": "value:file",
    },
  ],
  "sourceLocation": {
    "columnNumber": 2,
    "fileName": "fileName:file",
    "lineNumber": 1,
  },
  "stage": "running",
  "start": 1,
  "status": "unknown",
  "statusDetails": {
    "message": "message:file",
    "trace": "trace:file",
  },
  "stop": 102,
  "transformedCode": "transformedCode:file",
  "workerId": "2",
}
`;

exports[`mergeTestFileMetadata globalAndTestFile 1`] = `
{
  "attachments": [
    {
      "name": "global.txt",
      "source": "/tmp/global.txt",
      "type": "text/plain",
    },
    {
      "name": "file_docblock.txt",
      "source": "/tmp/file_docblock.txt",
      "type": "text/plain",
    },
    {
      "name": "file.txt",
      "source": "/tmp/file.txt",
      "type": "text/plain",
    },
  ],
  "currentStep": undefined,
  "description": [
    "description:global",
    "description:file_docblock",
    "description:file",
  ],
  "descriptionHtml": [
    "descriptionHtml:global",
    "descriptionHtml:file_docblock",
    "descriptionHtml:file",
  ],
  "displayName": "displayName:file",
  "fullName": "fullName:file",
  "historyId": "historyId:file",
  "labels": [
    {
      "name": "tag",
      "value": "global",
    },
    {
      "name": "tag",
      "value": "file_docblock",
    },
    {
      "name": "tag",
      "value": "file",
    },
  ],
  "links": [
    {
      "name": "link",
      "type": "global",
      "url": "https://example.com/global",
    },
    {
      "name": "link",
      "type": "file_docblock",
      "url": "https://example.com/file_docblock",
    },
    {
      "name": "link",
      "type": "file",
      "url": "https://example.com/file",
    },
  ],
  "parameters": [
    {
      "name": "parameter:global",
      "value": "value:global",
    },
    {
      "name": "parameter:file_docblock",
      "value": "value:file_docblock",
    },
    {
      "name": "parameter:file",
      "value": "value:file",
    },
  ],
  "sourceLocation": {
    "columnNumber": 2,
    "fileName": "fileName:file",
    "lineNumber": 1,
  },
  "stage": "scheduled",
  "start": 0,
  "status": "unknown",
  "statusDetails": {
    "message": "message:file",
    "trace": "trace:file",
  },
  "stop": 102,
  "transformedCode": "transformedCode:file",
  "workerId": "2",
}
`;

exports[`mergeTestFileMetadata merge 1`] = `{}`;

exports[`mergeTestFileMetadata merge 2`] = `
{
  "attachments": undefined,
  "currentStep": undefined,
  "description": undefined,
  "descriptionHtml": undefined,
  "displayName": undefined,
  "fullName": undefined,
  "historyId": undefined,
  "labels": undefined,
  "links": undefined,
  "parameters": undefined,
  "sourceLocation": undefined,
  "stage": undefined,
  "start": undefined,
  "status": undefined,
  "statusDetails": undefined,
  "stop": undefined,
  "transformedCode": undefined,
  "workerId": undefined,
}
`;

exports[`mergeTestFileMetadata steps 1`] = `
[
  {
    "attachments": [
      {
        "name": "step_invocation.txt",
        "source": "/tmp/step_invocation.txt",
        "type": "text/plain",
      },
    ],
    "currentStep": undefined,
    "displayName": "displayName:step_invocation",
    "hookType": undefined,
    "parameters": [
      {
        "name": "parameter:step_invocation",
        "value": "value:step_invocation",
      },
    ],
    "sourceLocation": {
      "columnNumber": 2,
      "fileName": "fileName:step_invocation",
      "lineNumber": 1,
    },
    "stage": "scheduled",
    "start": 10,
    "status": "passed",
    "statusDetails": {
      "message": "message:step_invocation",
      "trace": "trace:step_invocation",
    },
    "steps": [
      {
        "attachments": undefined,
        "currentStep": undefined,
        "displayName": undefined,
        "hookType": undefined,
        "parameters": undefined,
        "sourceLocation": undefined,
        "stage": undefined,
        "start": undefined,
        "status": undefined,
        "statusDetails": undefined,
        "steps": undefined,
        "stop": undefined,
        "transformedCode": undefined,
      },
    ],
    "stop": 110,
    "transformedCode": "transformedCode:step_invocation",
  },
]
`;
